### PR TITLE
bug 1490727: Recurrent payment force GitHub redirect

### PR DESCRIPTION
**Pull request by joshJarr**
**In this PR**
https://trello.com/c/SQgMeAXd/11-as-an-anonymous-user-i-want-to-learn-why-i-have-to-log-in-to-github-so-that-i-dont-leave-the-process

** notes for QA **
- [ ] ensure that the recurring payment form cannot be submitted until users login
- [ ] ensure one-time payment form still works for both anonymous and authed users
- [ ] ensure recurring payment form redirects to github and back to the FAQ page
- [ ] ensure form data is preserved

**Screenshot**
screenshot 2018-11-08 at 18 26 08

### CODE REVIEW

**Comment by schalkneethling** [comment]
* [x] Clicking the pay button while not logged in, shows the "We noticed your are not logged in...." messaging

**Comment by schalkneethling** [comment]
I do wonder whether we should not simply start of with the below screen if we know a user is not logged in but, want to make a recurring payment?

screenshot 2018-11-12 at 14 35 35

It feels annoying that, after having completed the form you are then road blocked by the login. The main reason is because, at this stage, the user has no idea that the information they have filled in will be preserved. Might cause some users to abandon at this stage.

If we do not want to change the flow to start with the above screen, perhaps we can add some copy to reassure the user that the info they entered will be preserved?

Thoughts @joshJarr @jwhitlock @atopal

**Comment by schalkneethling** [comment]
* [x] ensure recurring payment form redirects to github and back to the FAQ page
* [x] ensure form data is preserved

**Comment by schalkneethling** [comment]
* [x] Once off payment still works for non-authenticated users
* [x] Once off and recurring payments work as expected for logged in users with relevant for data pre-filled.

File: kuma/static/js/payments-handler.js [hunk 1]
```diff
@@ -458,7 +467,23 @@
...
+    /**
+     * Builds correct URL param and directs to GitHub for authentication.
+     * @param {jQuery.Event} event Event object.
+     */
+    function redirectUserToLogin(event) {
+        event.preventDefault();
+        var gitHubLink = $(this).attr('href');
+        var gitHubNext = $(this).data('next');
+        var getFormFields = form.serialize();
+        gitHubLink += '&next='+ gitHubNext + '?' + encodeURIComponent(getFormFields);
```
Comment (schalkneethling) [inline kuma/static/js/payments-handler.js]: Nit: space after '&next='

```suggestion
        gitHubLink += '&next=' + gitHubNext + '?' + encodeURIComponent(getFormFields);
```

File: kuma/payments/jinja2/payments/includes/payments-banner.html [hunk 2]
```diff
@@ -53,26 +56,33 @@ <h4 tabindex="-1">{{_('Support MDN, and it comes back to you.')}}</h4>
...
-                                    {#- include "payments/includes/payments-form.html" #}
+                                    {%  include "payments/includes/payments-form.html" %}
                                 {% endwith %}
                             {% else %}
-                                {% set github_url = provider_login_url('github', next=next_url) %}
-                                <p>{{_('We’ve noticed you’re not logged in to MDN. To complete your monthly payment, we’ll need you to log in first.')}}</p>
-                                <a href="{{ github_url }}" class="payment-form-button" data-service="GitHub" rel="nofollow">
-                                    {{ _('Sign in') }}{% include 'includes/icons/social/github.svg' %}
-                                </a>
+                                {% with form=form, hide_submit=True, is_popover=False, recurring_payment=recurring_payment, payments_url=payments_url, recurring_payment_subscription_url=recurring_payment_subscription_url  %}
```
Comment (schalkneethling) [inline kuma/payments/jinja2/payments/includes/payments-banner.html]: @joshJarr What is `hide_submit=True` used for? I do not see it being used anywhere else other than this line.

Comment (joshJarr) [reply kuma/payments/jinja2/payments/includes/payments-banner.html]: Addressed, great spot, I think this was leftover from a previous implementation. Thanks!

**Comment by CHHarding1989** [comment]
> I do wonder whether we should not simply start of with the below screen if we know a user is not logged in but, want to make a recurring payment?
>
> It feels annoying that, after having completed the form you are then road blocked by the login. The main reason is because, at this stage, the user has no idea that the information they have filled in will be preserved. Might cause some users to abandon at this stage.

Hey @schalkneethling, yup definitely a risk but I think its equally risky putting a log in button as the first interaction a user takes on the page — "I thought you wanted me to pay not log in?" they might say. Putting the log in step _after_ the form places auth in the wider process of completing a payment, similar to most ecommerce flows.

> If we do not want to change the flow to start with the above screen, perhaps we can add some copy to reassure the user that the info they entered will be preserved?

Yup could explore this. They only issue is that If the user doesn't have a Github account, we can't preserve their data after going through the Github sign up process. And at this step (prior to clicking 'sign in') we cant differentiate between users that have Github accounts and aren't logged in (who's data we can preserve) and users that flat out don't have a Github account (who's data we can't preserve.) So any copy reassuring that we could preserve data might be a lie to some users! 😬

**Comment by schalkneethling** [comment]
@CHHarding1989 Yup, that does make it tricky. Thanks for your feedback.

**Comment by schalkneethling** [comment]
Travis is not happy but, it seems all pull requests are failing on Travis because of:

```
ERROR: Service 'web' failed to build: The command '/bin/sh -c set -ex   && for key in     94AE36675C464D64BAFA68DD7434390BDBE9B9C5     FD3A5288F042B6850C66B31F09FE44734EB7990E     71DCFD284A79C3B38668286BC97EC7A07EDE3FC1     DD8F2338BAE7501E3DD5AC78C273792F7D83545D     C4F0DFFF4E8C1A8236409D08E73BC641CC11F4C8     B9AE9905FFD7803F25714661B63B535A4C206CA9     56730D5401028683275BD23C23EFEFE93C4CFFFE     77984A986EBC2AA786BC0F66B01FBB92821C587A     8FCCA13FEF1D0C2E91008E09770F7A9A5AE15600   ; do     gpg --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$key" ||     gpg --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$key" ||     gpg --keyserver hkp://pgp.mit.edu:80 --recv-keys "$key" ;   done' returned a non-zero code: 2
ERROR: InvocationError: '/usr/local/bin/docker-compose build web'
___________________________________ summary ____________________________________
ERROR:   docker: commands failed
The command "tox -v" exited with 1.
```

**Comment by jwhitlock** [comment]
A custom donation amount isn't saved for me in local development.

1. In local development admin, deleted my social account so I'd have to go through login
2. Went to http://localhost:8000/de/payments/recurring in private browsing session
3. Filled in "Bob Smith", "bob@example.com", custom amount 2.7, agreed to terms
4. Clicked "Pay $2.7", next was "Sign in" button
5. Next is GitHub login, 2fA
6. Back to MDN to create profile: jwhitlock_20181112, my email, etc.
7. Back to http://localhost:8000/de/payments/recurring/subscription?name=Bob+Smith&email=bob%40example.com&donation_amount=2.7&accept_checkbox=on:
    - ✅ Name is Bob Smith
    - ✅ Email is bob@example.com
    - ❌ Donation is $64/mo
    - ✅ Acceptance is checked

When I tried again with a pre-selected amount, $16/mo was preserved through the login. It appears to be custom amounts that are dropped and become $64/mo.

**Pull request merged**